{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c656352f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import re\n",
    "from tqdm.notebook import tqdm\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "633076fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>chromosome_path</th>\n",
       "      <th>plasmid_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RS_GCF_010645065.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GB_GCA_036004205.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GB_GCA_934196075.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GB_GCA_018674025.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GB_GCA_030430655.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140277</th>\n",
       "      <td>GB_GCA_030668065.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140278</th>\n",
       "      <td>GB_GCA_009889605.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140279</th>\n",
       "      <td>GB_GCA_041305945.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140280</th>\n",
       "      <td>GB_GCA_030699205.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140281</th>\n",
       "      <td>GB_GCA_022729255.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>140282 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 accession                                    chromosome_path  \\\n",
       "0       RS_GCF_010645065.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "1       GB_GCA_036004205.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "2       GB_GCA_934196075.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "3       GB_GCA_018674025.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "4       GB_GCA_030430655.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "...                    ...                                                ...   \n",
       "140277  GB_GCA_030668065.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "140278  GB_GCA_009889605.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "140279  GB_GCA_041305945.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "140280  GB_GCA_030699205.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "140281  GB_GCA_022729255.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "\n",
       "       plasmid_path  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2               NaN  \n",
       "3               NaN  \n",
       "4               NaN  \n",
       "...             ...  \n",
       "140277          NaN  \n",
       "140278          NaN  \n",
       "140279          NaN  \n",
       "140280          NaN  \n",
       "140281          NaN  \n",
       "\n",
       "[140282 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "split_summary_path = \"../out/2025-12-15_15-08-48/split_summary.csv\"\n",
    "df = pd.read_csv(split_summary_path)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3b9e8f4",
   "metadata": {},
   "source": [
    "2025-12-21 (Modify)\n",
    "- Add reason column to split_summary.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1201ad5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_reason_from_header(file_path):\n",
    "    \"\"\"\n",
    "    Read .fna.gz in file_path & Extract [class_reason=...]\n",
    "    \"\"\"\n",
    "    if pd.isna(file_path): # No plasmids\n",
    "        return None\n",
    "    \n",
    "    reasons = set()\n",
    "    try:\n",
    "        with gzip.open(file_path, 'rt') as f:\n",
    "            for line in f:\n",
    "                if line.startswith(\">\"):\n",
    "                    match = re.search(r\"\\[class_reason=(.*?)\\]\", line)\n",
    "                    if match:\n",
    "                        reasons.add(match.group(1))\n",
    "    except Exception as e:\n",
    "        return f\"error: {e}\"\n",
    "    \n",
    "    if not reasons:\n",
    "        return \"unknown\"\n",
    "    return \";\".join(sorted(reasons)) # for mutiple contigs & coressponding reasons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9de23329",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting plasmid reasons...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "40fba3f269fa42d2916764322908d861",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/140282 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting chromosome reasons...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9d59b167f4ed46eab3235312373dd833",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/140282 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>chromosome_path</th>\n",
       "      <th>plasmid_path</th>\n",
       "      <th>plasmid_reason</th>\n",
       "      <th>chromosome_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RS_GCF_010645065.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>no_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GB_GCA_036004205.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>no_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GB_GCA_934196075.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>no_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GB_GCA_018674025.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>no_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GB_GCA_030430655.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>no_match</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            accession                                    chromosome_path  \\\n",
       "0  RS_GCF_010645065.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "1  GB_GCA_036004205.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "2  GB_GCA_934196075.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "3  GB_GCA_018674025.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "4  GB_GCA_030430655.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "\n",
       "  plasmid_path plasmid_reason chromosome_reason  \n",
       "0          NaN           None          no_match  \n",
       "1          NaN           None          no_match  \n",
       "2          NaN           None          no_match  \n",
       "3          NaN           None          no_match  \n",
       "4          NaN           None          no_match  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Rows with plasmids ---\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>chromosome_path</th>\n",
       "      <th>plasmid_path</th>\n",
       "      <th>plasmid_reason</th>\n",
       "      <th>chromosome_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>RS_GCF_040957575.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>match=plasmid</td>\n",
       "      <td>match=chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>RS_GCF_033052795.2</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>match=plasmid</td>\n",
       "      <td>no_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>RS_GCF_011045815.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>match=plasmid</td>\n",
       "      <td>match=chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>RS_GCF_036964665.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>match=plasmid</td>\n",
       "      <td>no_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>GB_GCA_002813895.1</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>/home/yasutake/research/projects/metagenome/da...</td>\n",
       "      <td>match=plasmid</td>\n",
       "      <td>match=chromosome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              accession                                    chromosome_path  \\\n",
       "35   RS_GCF_040957575.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "43   RS_GCF_033052795.2  /home/yasutake/research/projects/metagenome/da...   \n",
       "238  RS_GCF_011045815.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "308  RS_GCF_036964665.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "336  GB_GCA_002813895.1  /home/yasutake/research/projects/metagenome/da...   \n",
       "\n",
       "                                          plasmid_path plasmid_reason  \\\n",
       "35   /home/yasutake/research/projects/metagenome/da...  match=plasmid   \n",
       "43   /home/yasutake/research/projects/metagenome/da...  match=plasmid   \n",
       "238  /home/yasutake/research/projects/metagenome/da...  match=plasmid   \n",
       "308  /home/yasutake/research/projects/metagenome/da...  match=plasmid   \n",
       "336  /home/yasutake/research/projects/metagenome/da...  match=plasmid   \n",
       "\n",
       "    chromosome_reason  \n",
       "35   match=chromosome  \n",
       "43           no_match  \n",
       "238  match=chromosome  \n",
       "308          no_match  \n",
       "336  match=chromosome  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved detailed summary to ../out/split_summary_reasons.csv\n"
     ]
    }
   ],
   "source": [
    "tqdm.pandas()\n",
    "\n",
    "print(\"Extracting plasmid reasons...\")\n",
    "df['plasmid_reason'] = df['plasmid_path'].progress_apply(extract_reason_from_header)\n",
    "\n",
    "print(\"Extracting chromosome reasons...\")\n",
    "df['chromosome_reason'] = df['chromosome_path'].progress_apply(extract_reason_from_header)\n",
    "\n",
    "# Confirm result\n",
    "display(df.head())\n",
    "\n",
    "print(\"\\n--- Rows with plasmids ---\")\n",
    "display(df[df['plasmid_path'].notna()].head())\n",
    "\n",
    "out_path = \"../out/2025-12-15_15-08-48/split_summary_reasons.csv\"\n",
    "df.to_csv(out_path, index=False)\n",
    "print(f\"Saved detailed summary to {out_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd809ebf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hyena",
   "language": "python",
   "name": "hyena"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
